### PR TITLE
Fix FS moves across Unix mounts

### LINKED ISSUE #13039: Move-Item does not work across mounts on Linux
## Summary

Just started using PowerShell core on Linux and its fantastic so far! I can across an interesting bug regarding the Move-Item cmdlet. If I'm on the same mount point, it works as expected but if I'm moving items from one Mount to another it errors with a cross device link error. 

Copy-Item is able to copy across devices as a work around. See below for details.

Hopefully this can get fixed in the 7.1 release.

## Steps to reproduce

```powershell
# Note: Source and destinations must be on different mounts shown by the "df" command.
$Source=/tmp/files/*
$Destination=/etc/files/
Move-Item $Source $Destination
```

## Expected behavior

```none
Files are removed from the source and exist at the destiation.
```

## Actual behavior

```none
Red Error Message:
    Move-Item: Invalid cross-device link
```

## Environment data

```none
$PSVersionTable:

Name                           Value
----                           -----
PSVersion                      7.0.2
PSEdition                      Core
GitCommitId                    7.0.2
OS                             Linux 3.10.0-1127.13.1.el7.x86_64 #1 SMP Tue Jun 23 15:46:38 UTC 2020
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

Other System Information:
Operating System: CentOS Linux release 7.8.2003 (Core)
UNAME: Linux op2-snf-ldev4.cgifederal.com 3.10.0-1127.13.1.el7.x86_64 #1 SMP Tue Jun 23 15:46:38 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux


System Mounts:
Filesystem                                Type     1K-blocks     Used Available Use% Mounted on
devtmpfs                                 devtmpfs   3987736        0   3987736   0% /dev
tmpfs                                       tmpfs      4004528        0   4004528   0% /dev/shm
tmpfs                                       tmpfs      4004528     9768   3994760   1% /run
tmpfs                                       tmpfs      4004528        0   4004528   0% /sys/fs/cgroup
/dev/mapper/centos_root       xfs        5232640   181832   5050808   4% /
/dev/mapper/centos_usr         xfs        8378368  4752188   3626180  57% /usr
/dev/sda1                                xfs        1038336   298172    740164  29% /boot
/dev/mapper/centos_tmp        xfs        8378368   183096   8195272   3% /tmp
/dev/mapper/centos_home     xfs      157209600   185644 157023956   1% /home
/dev/mapper/centos_var         xfs        8378368  3279240   5099128  40% /var
/dev/mapper/centos_opt         xfs      209612800 16759668 192853132   8% /opt
/dev/mapper/centos_var--lib  xfs        8378368   428752   7949616   6% /var/lib
tmpfs                                       tmpfs       800908       12    800896   1% /run/user/42
tmpfs                                       tmpfs       800908        0    800908   0% /run/user/1391052081

Note that LVM is used on this system, not sure if that's causing cross-device linking issues or not.

```

## Current Work Around

I am able to Copy-Item and then Remove-Item. This only impacts Move-Item for some reason.

```powershell
Copy-Item $Source $Destination -Recurse
Remove-Item $Source
```

### PR DESCRIPTION
**Pull request by rjmholt**
# PR Summary

Fixes https://github.com/PowerShell/PowerShell/issues/13039.

Currently `Move-Item` does not support cross-mount moves on Unix, because it doesn't check properly whether given paths lie on the same filesystem.

This PR fixes the check to identify when paths lie on different devices on Unix, so that the move logic can properly copy and delete rather than trying to call the `rename` API.

**NOTE** I'm not sure how we can test this, but I would love to if possible.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by rjmholt** [comment]
Ugh Visual Studio applied some autoformatting...

**Comment by KalleOlaviNiemitalo** [comment]
Would it be reasonable to try the `rename` call and recognize the `EXDEV` error? [rename(2)](https://linux.die.net/man/2/rename) says `rename` will fail with that if you try to rename between two mount points of the same file system; in which case, I think `st_rdev` will be equal.

**Comment by rjmholt** [comment]
> Would it be reasonable to try the rename call and recognize the EXDEV error?

We don’t call that directly, it’s a couple of layers down within .NET.

It turns out there’s a similar check done for Windows already with the `Move-Item` call and I think the best thing is to do something similar here.

PowerShell already has its own `stat` call, so basically we use that to check whether the device ID of the source and destination are the same.

Implementing that conforms with the existing logic and minimises the code changes required to fix this bug (which we may not be able to test).

**Comment by KalleOlaviNiemitalo** [comment]
[A comment in GNU coreutils](https://github.com/coreutils/coreutils/blob/439741053256618eb651e6d43919df29625b8714/src/mv.c#L212-L216) notes that comparing `st_dev` to decide whether to `rename` is not reliable with NFS.

**Comment by rjmholt** [comment]
> A comment in GNU coreutils notes that comparing st_dev to decide whether to rename is not reliable with NFS.

Hmmm, yeah that's good information.

I'm also interested by the fact that .NET makes file copying work across drives but not directory copying, making me feel that this might be unintended behaviour. Since their layer is better placed to deal with this, I'll follow up with them before we proceed.

**Comment by rjmholt** [comment]
Given the discussion in https://github.com/dotnet/runtime/issues/31149, I don't think .NET is about to fix this, so continuing with a workaround.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 1]
```diff
@@ -6072,6 +6063,42 @@ protected override bool IsItemContainer(string path)
...
+            {
+                // Rather than try to ascertain whether we can rename a directory ahead of time,
+                // it's both faster and more correct to try to rename it and fall back to copy/deleting it
+                // See also: https://github.com/coreutils/coreutils/blob/439741053256618eb651e6d43919df29625b8714/src/mv.c#L212-L216
+                CopyAndDelete(directory, destinationPath, force);
+            }
+#else
+            // On Windows, being able to rename vs copy/delete a file
+            // is just a question of the drive
+            if (IsSameWindowsVolume(directory.FullName, destinationPath))
```
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:6091]: .Net uses MOVEFILE_COPY_ALLOWED flag for MoveFileEx() on Windows.
So I think we can remove the check at all on all platforms but only check `catch (IOException e) when (e.HResult == UNIX_ERRNO_EXDEV)` on Unix-s.

Another thought is that we can a concern about performance (rename on the same volume is very fast but cross-volume one is a coping that can take many time.) and then we need to add a new parameter `Move-Item -AllowCrossVolumeCoping` to explicitly allow the behavior for all platforms. It is my preference. But how we will detect cross-volume on Windows if root path can be the same?

Comment (rjmholt) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:6091]: > .Net uses MOVEFILE_COPY_ALLOWED flag for MoveFileEx() on Windows.

That's very interesting. I did a small experiment with PS 7.1 and got this result though:

```
C:\Users\Robert Holt\Documents\Dev\Microsoft\PSScriptAnalyzer [implement-script-rules +4 ~9 -0 !]
> gi .\tools_copy\ | Sv t
C:\Users\Robert Holt\Documents\Dev\Microsoft\PSScriptAnalyzer [implement-script-rules +4 ~9 -0 !]
> $t.MoveTo('D:\tools')
MethodInvocationException: Exception calling "MoveTo" with "1" argument(s): "Source and destination path must have identical roots. Move will not work across volumes."
C:\Users\Robert Holt\Documents\Dev\Microsoft\PSScriptAnalyzer [implement-script-rules +4 ~9 -0 !]
> get-error

Exception             :
    Type           : System.Management.Automation.MethodInvocationException
    ErrorRecord    :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Exception calling "MoveTo" with "1" argument(s): "Source and destination path
must have identical roots. Move will not work across volumes."
            HResult : -2146233087
        CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : IOException
        InvocationInfo        :
            ScriptLineNumber : 1
            OffsetInLine     : 1
            HistoryId        : -1
            Line             : $t.MoveTo('D:\tools')
            PositionMessage  : At line:1 char:1
                               + $t.MoveTo('D:\tools')
                               + ~~~~~~~~~~~~~~~~~~~~~
            CommandOrigin    : Internal
        ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
    TargetSite     :
        Name          : ConvertToMethodInvocationException
        DeclaringType : System.Management.Automation.ExceptionHandlingOps,
System.Management.Automation, Version=7.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
        MemberType    : Method
        Module        : System.Management.Automation.dll
    StackTrace     :
   at System.Management.Automation.ExceptionHandlingOps.ConvertToMethodInvocationException(Exception
exception, Type typeToThrow, String methodName, Int32 numArgs, MemberInfo memberInfo)
   at CallSite.Target(Closure , CallSite , Object , String )
   at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0 arg0, T1 arg1)
   at System.Management.Automation.Interpreter.DynamicInstruction`3.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame
frame)
    Message        : Exception calling "MoveTo" with "1" argument(s): "Source and destination path
must have identical roots. Move will not work across volumes."
    Data           : System.Collections.ListDictionaryInternal
    InnerException :
        Type       : System.IO.IOException
        TargetSite :
            Name          : MoveTo
            DeclaringType : System.IO.DirectoryInfo
            MemberType    : Method
            Module        : System.IO.FileSystem.dll
        StackTrace :
   at System.IO.DirectoryInfo.MoveTo(String destDirName)
   at CallSite.Target(Closure , CallSite , Object , String )
        Message    : Source and destination path must have identical roots. Move will not work
across volumes.
        Source     : System.IO.FileSystem
        HResult    : -2146232800
    Source         : System.Management.Automation
    HResult        : -2146233087
CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
FullyQualifiedErrorId : IOException
InvocationInfo        :
    ScriptLineNumber : 1
    OffsetInLine     : 1
    HistoryId        : -1
    Line             : $t.MoveTo('D:\tools')
    PositionMessage  : At line:1 char:1
                       + $t.MoveTo('D:\tools')
                       + ~~~~~~~~~~~~~~~~~~~~~
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
```

> Another thought is that we can a concern about performance 

Personally I think user expectation will be for `Move-Item` to transparently deal with cross-volume moves without needing to specify a flag. `mv` doesn't require one and nor does .NET on Windows. Throwing an error is unhelpful and technically not performant -- it performs no useful work. I think the best way to get performance here is probably to optimise your mount layout.

**Comment by KalleOlaviNiemitalo** [comment]
Are there already tests for these scenarios:

- Move a directory to a subdirectory of itself via a symbolic link

  1. `a/b/c/d` is a file.
  3. `e` is a symbolic link to `a/b`.
  3. `Move-Item a e/f` must fail because that would move `a` to `a/b/f`, and the directory cannot be moved to a subdirectory of itself.
  4. After `Move-Item` has failed, `a/b/c/d` must still exist.

- Move a mount point

  1. `a` is a directory with a file system mounted on it. Thus, the directory cannot be deleted before the file system is unmounted.
  2. `a/b` is a file.
  3. `c` is an empty directory.
  4. `Move-Item a c/d` may succeed or fail.
  5. After `Move-Item` has succeeded, `c/d/b` must exist.
  6. After `Move-Item` has failed, either `c/d/b` or `a/b` must exist.

  Another program used to have a [bug](https://bugs.debian.org/166369) where it detected the inability to delete `a` and incorrectly tried to roll back by deleting `c/d` as well, losing all the files.

**Comment by rjmholt** [comment]
> Are there already tests for these scenarios:

I'm not familiar with the `Move-Item` tests, but I suspect all the tests we have are in the file I've added to here -- so if there's no test for a given scenario in there, there may be no test.

I'd prefer to keep discussion in this PR scoped, but your contribution of either an issue enumerating those tests or test code itself would be very welcome.

**Comment by rjmholt** [comment]
> Move a mount point

I'm not currently sure how we would reliably create mount points in our testing infrastructure, especially on the various platforms. I may need help from @adityapatwardhan and @JamesWTruher for that.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 2]
```diff
@@ -51,6 +51,12 @@ public sealed partial class FileSystemProvider : NavigationCmdletProvider,
                                                      ISecurityDescriptorCmdletProvider,
                                                      ICmdletProviderSupportsHelp
     {
+#if UNIX
+        // This is the errno returned by the rename() syscall
+        // when an item is attempted to be renamed across filesystem mount boundaries.
+        private const int UNIX_ERRNO_EXDEV = 18;
```
Comment (KalleOlaviNiemitalo) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:57]: 18 appears to be correct for Linux, macOS, and FreeBSD.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:6091]: If I remember correctly Windows Explorer on XP warned about cross-volume coping but later (Windows 7?) the behavior was changed. Perhaps we could follow this. 
Although we should take into account:
1. Performance - coping will be slow which may be unexpected for the user in an interactive session or it is amazing to see a script slowdown.
2. Lost security descriptor - after coping permissions are inherited from new root.
3. Formally it is a breaking change.
4. Reliability - it is not an atom operation.

Comment (madisonjsmith) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:6091]: "user expectation will be for Move-Item to transparently deal with cross-volume moves without needing to specify a flag"

I can't speak for all users but I can confirm that this is what I would expect as a user. :-)

Regarding *nix specific flags, a bit more context of  my scenario....The real value of running PowerShell (my opinion of course) on Linux is no longer having to maintain two separate shell script versions (bash and command prompt). 

If we start setting a precedence of having to use special flags for Linux or Microsoft or even differently named cmdlets, (Move-LinuxFile Move-MSFile) then as consumers, we likely have to add in  if ($Linux) {#Linux ways} else {#Microsoft ways} at which point we might as maintain separate shell scripts.

For the record, throwing errors should always be avoided from a user frustration standpoint but I would be ok if an error was displayed by default and the CopyAndDelete would happen when -Force is provided. 

With that said I like @rjmholt's solution and from one dev to another, I appreciate everyone's effort on this one!

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 3]
```diff
@@ -6072,6 +6063,42 @@ protected override bool IsItemContainer(string path)
...
+        /// </summary>
+        /// <param name="directory">The directory to move.</param>
+        /// <param name="destinationPath">The destination path to move the directory to.</param>
+        /// <param name="force">If true, force move the directory, overwriting anything at the destination.</param>
+        private void MoveDirectoryInfoUnchecked(DirectoryInfo directory, string destinationPath, bool force)
+        {
+#if UNIX
+            try
+            {
+                directory.MoveTo(destinationPath);
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:6079]: I would recommend a test hook here to throw the exception with HResult == UNIX_ERRNO_EXDEV.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs (hunk 2 above)
Comment (adityapatwardhan) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:57]: We should check the error code is 18 for all the variants we support especially Raspbian and Alpine, which could have it a bit different.

Comment (rjmholt) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:57]: Confirmed on RPi:

```
pi@daedalus:~ $ errno -l | grep EXDEV
EXDEV 18 Invalid cross-device link
```

Comment (rjmholt) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:57]: Confirmed on Alpine:

```
/ # errno -l | grep EXDEV
EXDEV 18 Cross-device link
```

**Review by adityapatwardhan — APPROVED** [review]

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:6096]: @rjmholt Can you remove the check on Windows and check cross-volume move? I expect it is work. 
(Although the question is how this works for reparse points like OneDrive but it seems we don't detect this whatever).

**Pull request merged**